### PR TITLE
Amber Lynn - Edges | API-Muncher

### PR DESCRIPTION
**Pull request by griffifam**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did you go about exploring the Edamam API, and how did you try querying the API? | I compared json to slack api to notices how things were structured within hashes and arrays.
What is an API Wrapper? Why is it in `lib`? How would your project change if you needed to interact with more than one API (aka more than just the Edamam API)? | I'd have to create a new wrapper. And it is in lib because it follows its own conventions and is not a part of the db
Describe your API wrapper, the methods you created in it, and why you decided to create those methods/how they were helpful | The API wrapper is what packages the parsed json data and returns it to controller. The wrapper talks to API
What was an edge case or failure case test you wrote for your API Wrapper? What was a nominal case? | I have not written tests
How does VCR aid in testing an API? | VCR records new episodes (responses) and plays back old responses
What is the Heroku URL of your deployed application? | I have not deployed

### CODE REVIEW

File: lib/edamam_api_wrapper.rb [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+
+  def self.show_recipe(id)
+    url = BASE_URL + "?app_id=#{ID}" + "&app_key=#{KEY}" + "&r=" + RECIPE_URI + "#{id}"
+
+    response = HTTParty.get(url)
+    if response.success?
+      recipe = EdamamApiWrapper.build_a_recipe(response)
+      return recipe
+    end
```
Comment (droberts-sea) [inline lib/edamam_api_wrapper.rb:36]: You should probably have an `else` clause here, that either returns nil or raises an error of your own choosing (to be rescued in the controller).

File: lib/edamam_api_wrapper.rb [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  end
+
+
+  private
+
+  def self.id_from_uri(uri)
+    return uri.split("_")[1]
+  end
```
Comment (droberts-sea) [inline lib/edamam_api_wrapper.rb:44]: I like that you broke this logic out as a separate method - good organization.

File: lib/edamam_api_wrapper.rb [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        ingredientLines: api_params["recipe"]["ingredientLines"],
+        healthLabels: api_params["recipe"]["healthLabels"]
+      }
+    )
+  end
+
+  def self.build_a_recipe(api_params)
+    return Recipe.new(
+      api_params[0]["label"],
+      api_params[0]["image"],
```
Comment (droberts-sea) [inline lib/edamam_api_wrapper.rb:63]: This method and the previous one are very similar. Could you consolidate them somehow?

File: app/controllers/searches_controller.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  end
+
+  def index
+    #search results
+    @recipes = EdamamApiWrapper.search_recipes( params[:recipe] )
+  end
+
+  def show
+    @recipe = EdamamApiWrapper.show_recipe( params[:id] )
+  end
```
Comment (droberts-sea) [inline app/controllers/searches_controller.rb:15]: In these controller actions, you should check for an error response (empty array in `index`, `nil` or an exception in `show`, and let the user know what happened, for example using a flash message.

File: config/routes.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+
+  root 'searches#root'
+
+  get ':searches/index', to: 'searches#index', as: 'results'
+
+  get 'searches/:id', to: 'searches#show', as: 'recipe'
```
Comment (droberts-sea) [inline config/routes.rb:8]: Why is there a route parameter on line 6 - it looks like your search form always uses the word "search" here, and sends the search term via query params.

Could you have used `resources :searches, only: [:index, :show]` here?

**Comment by droberts-sea** [comment]
# API Muncher
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Core Requirements** |
Git hygiene | yes
Comprehension questions | yes
**General** |
Rails fundamentals (RESTful routing, use of named paths) | see inline
Semantic HTML | some - missing high-level elements (`<header>`, `<footer>`, `<main>`)
Errors are reported to the user | no - A search with no hits results in a blank page, and mangling the URI on the show page results in an exception rather than a polite error message. Getting a good handle on error handling is not just good for your users, it will also make your life easier as a developer.
API Wrapper to handle the API requests | yes
Controller testing | no
Lib testing | no
Search Functionality | yes
List Functionality | yes
Show individual item functionality | yes (missing details, but the page is there)
**Styling** |
List view shows 10 items at a time and/or has pagination | no
Attractive and usable UI | no
**API Features** |
The App attributes Edamam | no - This is really important! When you're designing code for money, using someone else's work without attribution is a good way to end up in legal trouble.
The VCR cassettes do not contain the API key | N/A
**External Resources** |
Link to deployed app on Heroku | no
**Overall** |


This is a good start. I am definitely disappointed not to see tests - the work around mocking the external dependency with VCR is an important learning goal for this assignment, and not something we'll have the opportunity to come back to. Make sure you understand how and why we do that. I would also like to see more of a focus on error handling, this is a small thing that makes a surprisingly big difference in your development as an engineer.


However, the main functionality of the site is there - searching and viewing details work great, and it is clear to me that the learning goals around working with an API and building non-model classes for a Rails app were met. It feels to me like most of the missing pieces are more a result of not having enough time than not understanding. Let me know if you disagree, or if there's specific places you'd like additional support, and keep up the hard work!
